Doc 9216 import data (Java)

osfameron opened the pull request:
This is VERY draft, and quite horrific, have included dev team to ask for pointers on better more Java way of doing this.

* I realise this is syncronous, can rewrite to async, but not sure how best to verify that it works robustly / faster / etc.
* getting muddled with types (thus 2 functions for key/process based on type of Map)
* anything else horrible that needs fixing...

osfameron committed:
DOC-9216: Import Data example code for guide

Example code for the guide at:
  https://github.com/couchbase/docs-server/pull/2317

osfameron commented:
NB, though @mikereiche gave a workaround for the non-generic Map in Slack (thanks!), I haven't figured out how to fold it in without obscuring the intent of the code/docs.

maria-robobug commented on modules/howtos/examples/Import.java:
```diff
@@ -0,0 +1,209 @@
+  
+  // tag::upsertDocument[]
+  public void upsertRow(Map row) {
+    
```
Extra line here.

osfameron committed:
DOC-9216: Import Data (batch)

batch for JSON so far

maria-robobug commented on modules/howtos/examples/Import.java:
```diff
@@ -0,0 +1,314 @@
+      }
+      System.out.println("DONE");
+    }
+    // end::importJSON_batch[]
```
Indentation looks a bit weird here.

osfameron replied:
ta! (copy/paste edit artefact, have tidied up in subsequent commit

maria-robobug commented on modules/howtos/examples/Import.java:
```diff
@@ -0,0 +1,314 @@
+    public String toString() {
+      return "JsonDocument{id='" + id + "', content=" + content + "}";
+    }
+  }
```
Think there's extra spacing here, should be the closing bracket so no space.

maria-robobug commented on modules/howtos/examples/Import.java:
```diff
@@ -0,0 +1,314 @@
+    results.blockLast(Duration.of(60, ChronoUnit.SECONDS));
+
+    // ...
+    // tag::repeated[]
```
Is there a section missing here, or should this tag be removed?

osfameron replied:
believe I've removed this (renamed to ::omit also, which is more generally useful)

maria-robobug reviewed (approved):
👍 LGTM, just some minor comments!

osfameron committed:
DOC-9216: Import Data - batch examples

 - Refactor `preprocess`
   - adding JsonDocument value class (from Maria's examples)
 - add Flux:: examples for each filetype
   - NB: these all require different conversion
     (Flux::generate, ::fromIterable, or ::using)
     which may be inelegant/confusing?

osfameron commented:
Updated with Reactor examples, @daschl @programmatix @dnault would really appreciate a once-over on this Java code as this is my first time use of Flux:: and various other libraries.

dnault commented on modules/howtos/examples/Import.java:
```diff
@@ -0,0 +1,411 @@
+  public void importCSV() {
+    try {
+      CSVReaderHeaderAware csv = new CSVReaderHeaderAware(
+        new FileReader("howtos/examples/import.csv"));
```
As a rule of thumb, code that creates a Reader or InputStream is also responsible for closing it. `try-with-resources` makes it easy:

    try (CSVReaderHeaderAware csv = new CSVReaderHeaderAware(
         new FileReader("howtos/examples/import.csv"))) {
    ...
    }

dnault commented on modules/howtos/examples/Import.java:
```diff
@@ -0,0 +1,411 @@
+    System.out.println("importCSV_batch");
+    // end::omit[]
+    
+    Flux<Map<String,String>> rows = Flux.generate(
```
Could use the Flux.generate() overload that takes a final cleanup callback to close the reader when the flux terminates. Just add a third argument: `state -> state.close()` (or if there's an IOException, catch and wrap in a RuntimeException)

https://projectreactor.io/docs/core/release/api/reactor/core/publisher/Flux.html#generate-java.util.concurrent.Callable-java.util.function.BiFunction-java.util.function.Consumer-

dnault commented on modules/howtos/examples/Import.java:
```diff
@@ -0,0 +1,411 @@
+          else { sink.next(row); }
+          return state;
+        }
+        catch (IOException e) { throw new Error("IOException"); }
```
Instead of catching different exceptions and re-throwing, how about signalling the error like:

    catch (Exception t) {
        sink.error(t);
    }

As an aside, `Error` is typically reserved for conditions that should never, *EVER* happen. Like, inconceivable situations that indicate something is fundamentally wrong with the code. If you just want to smuggle a checked exception, wrapping it in a `RuntimeException` is usually the way to go.

osfameron replied:
good idea! (I wasn't sure if I should name each exception to document which ones might happen? But happy to keep it simpler...)

osfameron replied:
hmm, sink.error gives me a Generics type error, but I've used `throw new RuntimeException(e)` elsewhere at your suggestion anyway, so may keep with that for consistency.

dnault commented on modules/howtos/examples/Import.java:
```diff
@@ -0,0 +1,411 @@
+      .flatMap(doc -> reactiveCollection.upsert(doc.getId(), doc.getContent()));
+      
+    results.subscribe(System.out::println);
+    results.blockLast(Duration.of(60, ChronoUnit.SECONDS));
```
NIT: `Duration.ofSeconds(60)`

osfameron replied:
oh, even better, this was in an example I, um, copy/pasted, and I did think `ChronoUnit` looked quite over-the-top

dnault commented on modules/howtos/examples/Import.java:
```diff
@@ -0,0 +1,411 @@
+      .map(row -> preprocess(row))
+      .flatMap(doc -> reactiveCollection.upsert(doc.getId(), doc.getContent()));
+      
+    results.subscribe(System.out::println);
```
This code is upserting the documents twice, since `subscribe` and `blockLast` both subscribe to the flux (triggering its execution).

Recommendation is to remove the `subscribe`. If you want to print each mutation result, one way would be to chain `.doOnNext(System.out::println)` after `flatMap`.

osfameron replied:
ah, I thought as it's push based, the code would be run once and the *result* sent to all subscribers?
I'll have to review my mental model... thanks, will look at .doOnNext!

dnault commented on modules/howtos/examples/Import.java:
```diff
@@ -0,0 +1,411 @@
+
+  
+  // tag::importTSV[]
+  public void importTSV() {
```
NIT: There's an opportunity to refactor, since this is the same as the CSV version but with a differently configured CSV parser. Repeating the code is probably fine the documentation perspective, but we'll need to be careful to keep all of these versions in sync.

osfameron replied:
Agreed, and yes, would like to refactor but risk of confusing docs intent.

dnault commented on modules/howtos/examples/Import.java:
```diff
@@ -0,0 +1,411 @@
+
+    try {
+      String content  = new String(
+        Files.readAllBytes(
```
NIT: Perhaps a note that this will fail if the whole file doesn't fit in memory? Ideally we'd want to use a JSON stream parser to read each object one by one, but maybe that's out of scope.

osfameron replied:
Good catch. Yes, ideally would be streaming, using the Jackson Streaming API... 

I think this might be for a future version though (will check priority with @simon-dew) unless you happen to have time to contribute the code.

dnault commented on modules/howtos/examples/Import.java:
```diff
@@ -0,0 +1,411 @@
+    System.out.println("importJSONL_batch");
+    // end::omit[]
+
+    Flux<String> lines = Flux.using(
```
NIT: Here's how Simon recommends doing this (see the "conclusion" section of https://simonbasle.github.io/2017/10/file-reading-in-reactor/ )

    Flux<String> lines = Flux.defer(() ->
            Flux.fromStream(Files.lines(thePath)));

osfameron replied:
Thanks - he uses the `.using` construct in the main body of the article though, and the `.defer` only in the footnoted caveat.

Also .using does some error handling which the .defer incantation doesn't so not sure we'd gain any clarity by switching here?

dnault reviewed (changes requested):
These are some great examples.

maria-robobug commented on modules/howtos/examples/Import.java:
```diff
@@ -0,0 +1,411 @@
+}
+
+// tag::JsonDocument[]
+class JsonDocument {
```
Hmm, I think the build is failing due to this class existing in https://github.com/couchbase/docs-sdk-java/blob/release/3.2/modules/hello-world/examples/KvBulkHelloWorld.java 

Because all these classes are part of the same "package", it's still accessible when using `class <Your-Class-Name-Here>`, so it comes up as a duplicate. 
I *think* `private class` could mitigate this, if you want to keep this in here. Or even rename it to something else.

Another option would be to reference the tag in `KvBulkHelloWorld` for the class, but it might look odd to the user if they don't see it when they access the full example on Github.

osfameron replied:
woo, lemme give private a go

osfameron replied:
(ended up nesting the class instead)

osfameron committed:
DOC-9216: Import Data tweaks

in response to code reviews from dnault and Maria

osfameron committed:
DOC-9216: Import Data additional review issues

Don't run subscribe, to avoid running the upsert twice in Reactor examples.

Brief mention that JSON parsing currently slurps whole file into memory
(this will also be surfaced in docs)

osfameron committed:
Merge branch 'release/3.2' into DOC-9216-import-data
